### ISSUE TITLE
"Dialog" is not announced for modals

### ISSUE THREAD

**Issue by Malvoz** [twbs/bootstrap#33715]
**Repro steps:**
1. Start NVDA (or perhaps the screen reader of your choice)
2. Visit https://getbootstrap.com/docs/5.0/components/modal/#live-demo
3. Open the modal dialog

Expected:
"Dialog" should be announced (along with the content of the modal).

Result:
"Dialog" is not announced.

I noticed that manually moving `role="dialog"` (although moving `tabindex="-1"` instead may also work) from its current `<div class="modal">` element to the child `<div class="modal-dialog">` the result is that "dialog" _is_ announced.

This may be an NVDA bug (I'll probably wait with reporting until I get feedback here). It could be that this has to do with the fact that the standard `<dialog>` element must not have a `tabindex` attribute, and therefore for some reason NVDA fails to announce the ARIA role, but perhaps that's totally unrelated.

If others can confirm that this does not work as expected (in NVDA and/or another screen reader) that'd be much appreciated.

**Environment:**

- Windows 10.0.19042
- Chrome 90.0.4430.85
- NVDA 2020.4

/cc @patrickhlauke

**Comment by patrickhlauke** [comment]
it feels like perhaps there's some race condition happening here. I see that the `role="dialog"` and `aria-modal="true"` are dynamically added/removed from modals. it could be that NVDA isn't catching up quickly enough when the focus is moved to the newly opened modal that the role for where it's now landed has been changed.

but yes confirming that NVDA does not announce the fact focus has shifted to a modal dialog. JAWS (which does have other quirks relating to putting focus in the wrong place, it seems) does announce it.

think the whole modal logic may need to be reviewed - also, the entire underlying page should really be `aria-hidden` to hide it from AT (JAWS still lets you navigate behind the overlay, as it doesn't care about `aria-modal` just yet)

**Comment by patrickhlauke** [comment]
ok, tested the theory about the race condition by having the `role` and `aria-modal` already on the modal container even when not shown yet, but that had no effect. so this seems partly a shortcoming of NVDA, but also likely caused by the how focus is set to the container itself rather than inside it. it's arguably not a showstopper that NVDA doesn't explicitly say you're in a modal provided that the underlying page is hidden (which chrome/nvda and firefox/nvda seem to do correctly), but still worth looking into.

**Comment by alpadev** [comment]
@patrickhlauke You are right about our focusing behavior. I made some tests and once you focus the element with `role="dialog"`, it will announce the heading without mentioning it's a dialog.

Not sure how @Malvoz changed the element with the `role` attribute, since we add this in our JS. If he did it by adding `role` in the dev-tools it would end up on both elements. _Duplicating_ `role="dialog"` to the child `<div class="modal-dialog">` causes NVDA to announce it's a dialog but doesn't announce the control buttons (which it usually would to my findings).

_Moving_ `role="dialog"` to `.modal-dialog` causes NVDA to announce the title only (likely because of the `aria-labelledby` attribute) without mentioning it's a dialog nor announce any controls.

IMO focusing the `.modal` element in it's current implementation isn't ideal anyway since that element is styled to be fixed and to span the whole viewport so you can't even see its focused.

**Comment by patrickhlauke** [comment]
Yes, focusing the whole of the dialog itself is now not a recommended approach anymore (especially for general dialogs that may have a lot of content), as this indeed can lead to the entirety of the dialog being just vomited out in one go by AT.

Related https://github.com/w3c/aria-practices/issues/442
